### PR TITLE
Allow ingesters to stay in the ring during restart

### LINKED ISSUE #2823: Store-gateway blocks resharding during rollout
When running the blocks storage, store-gateways reshard blocks whenever the ring topology changes. This means that during a rollout of the store-gateways (ie. deploy a config change or version upgrade) blocks are resharded across instances. 

This is highly inefficient in a cluster with a large number of tenants or few very large tenants. Ideally, no blocks resharding should occur during a rollout (if the blocks replication factor is > 1).

## Rollouts

We could improve the system to avoid the blocks resharding when the following conditions are met:
- `-experimental.store-gateway.replication-factor` is `> 1` (so that while a store-gateway restarts all of its blocks are replicated to at least another instance)
- `-experimental.store-gateway.tokens-file-path` is configured (so that previous tokens are picked up on restart)
- The store-gateway instance ID is stable across restarts (ie. Kubernetes Statfulsets)

To avoid blocks resharding during store-gateways rollout we need the restarting store-gateway instance to **not** be unregistered from the ring during the restart.

When a store-gateway shutdowns, the instance could be left in the `LEAVING` state within the ring and we could change the `BlocksReplicationStrategy.ShouldExtendReplicaSet()` to **not** extend the replica set if an instance is in the `LEAVING` state.

This means that during the rollout, for the blocks hold by the restarting replica there will be N-1 replicas (contrary to the N desired replicas configured). Once the instance restarts, it will have the same instance ID and same tokens (assuming `tokens-file-path` is configured) and thus will replace its state from `LEAVING` to `JOINING` within the ring.

## Scale down

There's no way to distinguish between a rollout and a scale down: the process just receives a termination signal.

This means that during a scale down, the instance would be left in the `LEAVING` state within the ring. However, the store-gateway has an auto-forget feature which removes unhealthy instances after 10x heartbeat timeouts (default: 1m timeout = 10m before an unhealthy instance is forgotten).

A scale down of a number of instance < replication factor could leverage on the auto-forget. However, there's no easy way to have a smooth scale down unless we'll have a way to signal the process whether it's going to shutdown because of a scale down or rollout.

## Crashes

_In case a store-gateway crashes, there would be no difference compared to today._

### PR DESCRIPTION
**Pull request by csmarchbanks**
By keeping an ingester in the ring and not writing to additional
ingesters while it is being restarted, will stop small chunks from being
persisted to other ingesters. This will allow users running with a WAL
to avoid excess load and memory during a rolling restart.

I am curious what others think of this, if accepted, I think a similar
approach could be applied to the store-gateway (issue #2823). The
downside is that scaling a cortex cluster down while running with skip
unregister will involve manual forgetting of the ingester (or automation
to do this).

Fixes: #3304

**Checklist**
- [x] Tests updated (some tests use old way, some use new way).
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md
Comment (csmarchbanks) [inline CHANGELOG.md:53]: Well, this is for myself, but may as well be consistent to make sure I don't forget :)

:wave: The v1.5 release just started and you will need to rebase to the latest master and move any CHANGELOG entries into the unreleased section. Thanks!

File: pkg/ring/lifecycler.go
Comment (pstibrany) [inline pkg/ring/lifecycler.go:59]: I'd suggest calling it UnregisterFromRing, and make it true by default (for compatibility).

File: pkg/ring/lifecycler.go
Comment (pstibrany) [inline pkg/ring/lifecycler.go:789]: This comment is slightly misleading, because by passing "true", we're actually disabling the unregister. (Why I suggested the rename)

File: pkg/ring/ring.go
Comment (pstibrany) [inline pkg/ring/ring.go:130]: We can extend the description to explain why would someone want to disable this, and that it's perhaps best used with disabled unregister?

**Review by pstibrany — COMMENTED** [review]
Thank you. I've left some comments/suggestions. Can we also modify api/_index.md to document that /shutdown will unregister from the ring, even when unregistering is disabled?

**Comment by csmarchbanks** [comment]
Thanks for the review! I implemented all of the suggestions - the rename did mean changing some tests around that were depending on SkipUnregister being default false.

**Review by pstibrany — APPROVED** [review]
LGTM, thanks. Let's wait for @pracucci's review as well, before merging, please.

**Comment by csmarchbanks** [comment]
Friendly reminder for @pracucci to review as requested :)

File: pkg/ring/lifecycler.go
Comment (jtlisi) [inline pkg/ring/lifecycler.go:784]: **suggestion:**

```suggestion
// UnregisterFromRing returns if unregistering is skipped on shutdown.
```

File: pkg/ring/lifecycler.go
Comment (jtlisi) [inline pkg/ring/lifecycler.go:785]: **suggestion(non-blocking,if-minor):** I would rename this function to parallel `SetUnregisterFromRing`. In my opinion `UnregisterFromRing()` connotes that it will actually perform the operation to unregister from the ring, not return the config value.

```suggestion
func (i *Lifecycler) GetUnregisterFromRing() bool {
```

**Review by jtlisi — APPROVED** [review]
LGTM, modulo some minor nits/suggestions

File: pkg/ring/lifecycler.go
Comment (csmarchbanks) [reply pkg/ring/lifecycler.go]: Do'h, I really thought I renamed that too... Fixed.

File: pkg/ring/lifecycler.go
Comment (csmarchbanks) [reply pkg/ring/lifecycler.go]: I changed it to `Should` which read nicer to me than `Get`.

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Comment by bboreham** [comment]
Ref #467 which I think this fixes, at least for the WAL restart-in-place case.

**Comment by bboreham** [comment]
I'll mention another downside: you lose redundancy.

With replication-factor=3, if you turn off "extend writes" then do a rolling update, you are down to 2 copies of some data. If one of those is then lost then queries will be unstable due to https://github.com/cortexproject/cortex/issues/731.

**Pull request merged**
